From 06cd4643b60a5da7e80514e1b2cb9f276142246b Mon Sep 17 00:00:00 2001
From: Paul Hunt <hunt@ti.com>
Date: Mon, 7 Mar 2011 17:16:14 -0700
Subject: syslink: ipu_pm: set ivahd to hw_auto mode

Change CM_IVAHD_CLKSTCTRL to HW_AUTO mode after
IVAHD gets started.

Change-Id: I38c4100a4d5f32a656b3b65af5557a41bcd55d36
Signed-off-by: Paul Hunt <hunt@ti.com>
---
 drivers/dsp/syslink/ipu_pm/ipu_pm.c |   30 ++++++++++++++++++++++++++++++
 1 files changed, 30 insertions(+), 0 deletions(-)

diff --git a/drivers/dsp/syslink/ipu_pm/ipu_pm.c b/drivers/dsp/syslink/ipu_pm/ipu_pm.c
index 9e33868..109de1a 100644
--- a/drivers/dsp/syslink/ipu_pm/ipu_pm.c
+++ b/drivers/dsp/syslink/ipu_pm/ipu_pm.c
@@ -33,6 +33,7 @@
 #include <linux/clk.h>
 #include <linux/uaccess.h>
 #include <linux/irq.h>
+#include <linux/timer.h>
 
 #include <linux/platform_device.h>
 #include <syslink/notify.h>
@@ -66,6 +67,9 @@
 /* Module headers */
 #include <plat/ipu_dev.h>
 #include "../../../../arch/arm/mach-omap2/cm.h"
+#include "../../../../arch/arm/mach-omap2/prm.h"
+#include "../../../../arch/arm/mach-omap2/prcm-common.h"
+#include "../../../../arch/arm/mach-omap2/prm44xx.h"
 #include "ipu_pm.h"
 
 /** ============================================================================
@@ -310,6 +314,9 @@ struct work_struct clean;
 static DECLARE_COMPLETION(ipu_clean_up_comp);
 static bool recover;
 
+/* kernel timers */
+struct timer_list ivahd_init_timer;
+
 /* Latency cstrs */
 #ifdef CONFIG_OMAP_PM
 static struct pm_qos_request_list *pm_qos_handle;
@@ -1380,6 +1387,21 @@ static inline int ipu_pm_get_ivaseq0(struct ipu_pm_object *handle,
 	return PM_SUCCESS;
 }
 
+void ivahd_hwauto_init(unsigned long arg)
+{
+	pr_debug("%s delayed IVAHD HW_AUTO enable activated\n", __func__);
+
+	/* set Next State to INACTIVE for IVAHD */
+	prm_write_mod_reg(0xff0e02,
+			 OMAP4430_PRM_IVAHD_MOD,
+			 OMAP4_PM_IVAHD_PWRSTCTRL_OFFSET);
+
+	/* do the delayed switch to HW_AUTO for IVAHD */
+	cm_write_mod_reg(HW_AUTO,
+			 OMAP4430_CM2_IVAHD_MOD,
+			 OMAP4_CM_IVAHD_CLKSTCTRL_OFFSET);
+}
+
 /*
   Request IVA SEQ1 on behalf of an IPU client
  *
@@ -1413,6 +1435,14 @@ static inline int ipu_pm_get_ivaseq1(struct ipu_pm_object *handle,
 	params->pm_sl2if_counter++;
 	pr_debug("Request sl2if\n");
 
+	/* Need to delay after startup to allow time
+	  * for MM to load code and start ivahd running
+	  */
+	init_timer(&ivahd_init_timer);
+	ivahd_init_timer.expires = jiffies + 100;
+	ivahd_init_timer.function = ivahd_hwauto_init;
+	add_timer(&ivahd_init_timer);
+
 	return PM_SUCCESS;
 }
 
-- 
1.6.0.4

